src/Networks/Network.ts

```diff
@@ -0,0 +1,323 @@
+            this.dhcpd_relay_enabled = props.dhcpd_relay_enabled;
+        }
+        if (!Validate.isUndefined(props.dhcpd_boot_enabled)) {
+            this.dhcpd_boot_enabled = props.dhcpd_boot_enabled;
+        }
+        if (!Validate.isUndefined(props.igmp_proxy_downstream)) {
+            this.igmp_proxy_downstream = props.igmp_proxy_downstream;
+        }
+        if (!Validate.isUndefined(props.upnp_lan_enabled)) {
+            this.upnp_lan_enabled = props.upnp_lan_enabled;
+        }
+        if (!Validate.isUndefined(props.dhcpd_ntp_enabled)) {
+            this.dhcpd_ntp_enabled = props.dhcpd_ntp_enabled;
+        }
+        if (!Validate.isUndefined(props.mdns_enabled)) {
+            this.mdns_enabled = props.mdns_enabled;
+        }
+        if (!Validate.isUndefined(props.lte_lan_enabled)) {
+            this.lte_lan_enabled = props.lte_lan_enabled;
+        }
+        if (!Validate.isUndefined(props.dhpcd_tftp_server)) {
+            this.dhpcd_tftp_server = props.dhpcd_tftp_server;
+        }
+        if (!Validate.isUndefined(props.auto_scale_enabled)) {
+            this.auto_scale_enabled = props.auto_scale_enabled;
+        }
+        if (!Validate.isUndefined(props.vlan)) {
+            this.vlan = props.vlan;
+        }
+        if (!Validate.isUndefined(props.l2tp_allow_weak_ciphers)) {
+            this.l2tp_allow_weak_ciphers = props.l2tp_allow_weak_ciphers;
+        }
+        if (!Validate.isUndefined(props.l2tp_local_wan_ip)) {
+            this.l2tp_local_wan_ip = props.l2tp_local_wan_ip;
+        }
+        if (!Validate.isUndefined(props.vpn_type)) {
+            this.vpn_type = props.vpn_type;
+        }
+        if (!Validate.isUndefined(props.radiusprofile_id)) {
+            this.radiusprofile_id = props.radiusprofile_id;
+        }
+        if (!Validate.isUndefined(props.dhpcd_wins_enabled)) {
+            this.dhpcd_wins_enabled = props.dhpcd_wins_enabled;
+        }
+        if (!Validate.isUndefined(props.require_mschapv2)) {
+            this.require_mschapv2 = props.require_mschapv2;
+        }
+        if (!Validate.isUndefined(props.l2tp_interface)) {
+            this.l2tp_interface = props.l2tp_interface;
+        }
+        if (!Validate.isUndefined(props.exposed_to_site_vpn)) {
+            this.exposed_to_site_vpn = props.exposed_to_site_vpn;
+        }
+        if (!Validate.isUndefined(props.x_ipsec_pre_shared_key)) {
+            this.x_ipsec_pre_shared_key = props.x_ipsec_pre_shared_key;
+        }
+
+        return this;
+    }
+
+    public async update(props: Partial<INetwork>): Promise<void> {
+        if (!Validate.isDefinedNotNull(props._id)) {
+            throw new Error('Cannot update network without _id');
+        }
+
+        return (
+            await this.instance.put('/rest/networkconf/:id', props, {
+                urlParams: {
+                    site: this.site.name,
+                    id: this._id
+                }
+            })
+        ).data?.data;
```

Comment:
on my side I change this to : 
```typescript
        await this.instance.put('/rest/networkconf/:id', props, {
            urlParams: {
                id: this._id
            }
        });

        this.import(props);
```

so, the current changes are refleted in the object . 
```typescript
    const networks = await site.networks.list();
    const network = networks.find((n) => n.name === 'test');

    if (!network) {
        throw new Error('not found');
    }

    console.log(network.enabled); // => true

    await network.update({ enabled: false });
    
    console.log(network.enabled); // => false
```

Reply:
Okay, I misunderstood the requirements there.  In your initial request you'd wanted update() to be able to update the Network without requiring querying. 

> like, I can, skip the loading step, and directly update a network if I know the correct _id

This was why I had the method return void.  If you allow a partial to be passed into update then perform `this.import(props);` at the end of the service call Network would be poorly formed with only partial information.

Now I could do what I'm doing in save() and take the return from UniFi which looks to be the entire Network object and import that.  That way update would be fully formed (at least until UniFi changes their API).

Thoughts?

Reply:
@KyleBrandon the problem here, is in my test, the update return nothing ... 

with this, it will update the current object ( so, when I do the list, the object if filled ... and when I update, the import will only update the updated props )

```typescript
const networks = await site.networks.list();
const network = networks.find((n) => n.name === 'test');


console.log(network);
/**
{
"name": "test",
...
"enabled": true
}
*/

await network.update({ enabled: false });

console.log(network);
/**
{
"name": "test",
...
"enabled": false
}
*/

```

Reply:
Fair point.  The one benefit of importing the properties returned by UniFi is that you would then have a consistent representation from UniFi if something was changed in another method (the UI for instance).

Reply:
yes yes . I agree ... the problem was, on my side, the update return an empty array ... it's different on your side ?

Reply:
No, that's correct, it was how I was implementing. I was waffling between having update() modify the Network state or not.  
Here's what I propose:

- I'll modify update() to import the result from UniFi, that way we have a consistent state with the gateway.  
- I also propose adding a Validate.isStringEmpty() to verify the id is not undefined, null or empty.  I could do isIdEmpty but I don't know all the validation rules around a UniFi network id.  If you don't like it on Validate I can add it to Network.

Does that sound reasonable?

Reply:
I'm not sure about the second part ... you want to pass an _id to update another network ? 

about the first one, yes if you can retrieve only the current network ? 

-- 
after some tests, I found : 
```
        const res = await this.instance.get('/rest/networkconf/:id', {
            urlParams: {
                id: this._id
            }
        });
```
so, we can call it to import ? 

something like : 
```
    public async load(): Promise<this> {
        const res = (
            await this.instance.get('/rest/networkconf/:id', {
                urlParams: {
                    id: this._id
                }
            })
        ).data;

        if (res.data) {
            this.import(res.data);
        }

        return this;
    }
```

( and so, you can call load at the end of the update ? )

Reply:
Currently we have these checks on the constructor, update(), and save().
```
        if (!Validate.isDefinedNotNull(this._id)) {
            throw new Error('Cannot update network without _id');
        }
```
I propose to do something like:
```
    isIdEmpty(): boolean {
        if (!Validate.isDefinedNotNull(this._id) || this._id.trim().length === 0) {
            return true;
        }

        return true;
    }
```

I would also add a check in update() to ensure that the props._id === this._id so someone couldn't update a different network as you indicated.

Kyle

Reply:
hum ... I'm not fan of the idea . 

I think, because the object need _id to be construct, other parts need to start with the fact that _id is always defined . 

So, maybe just something like : 
```typescript
    public async update(props: Partial<Omit<INetwork, '_id'>>): Promise<void> {
        await this.instance.put('/rest/networkconf/:id', props, {
```

Reply:
That works slick.

Reply:
Well, I misunderstood what I was seeing in the logs from Axios.  Turns out this conversation is somewhat moot as UniFi does _not_ return the current Network after a put.  I apologize for the misunderstanding, but I did find the debate valuable to understand your thoughts on the direction of the API.

Here are the current save() and update ():

```
    public async update(props: Partial<Omit<INetwork, '_id'>>): Promise<void> {
        await this.instance.put('/rest/networkconf/:id', props, {
            urlParams: {
                id: this._id
            }
        });
        this.import(props);
    }

    public async save(): Promise<this> {
        if (!Validate.isDefinedNotNull(this._id)) {
            throw new Error('Cannot update network without _id');
        }
        const network: INetwork = { ...this } as INetwork;
        await this.instance.put('/rest/networkconf/:id', network, {
            urlParams: {
                id: this._id
            }
        });
        return this;
    }
```

As and additional indication that _id is required and immutable I marked the property _readonly_ and set it in the constructor.

Reply:
no problem . 

About the readonly on _id . Yes, it can be a good idea :) . 

but remember that this checks is only typescript ... so, when a javascript user use it ... it can be overrided ( maybe with a getter/setter . or with a [javascript private class field](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields) ... but it is maybe too much for this time ) . 

( also, remember that readonly need to be defined in constructor, and not in the import ) . 


---

Another thing I saw, what about moving this.debug initialization to constructor ? 
something like : 
````typescript
const debug = createDebugger('Network');

export class Network extends _ObjectSubSite {
    constructor(config: IObjectSubSiteConfig, props: Partial<INetwork>) {
        super(config);

        if (!props._id) {
            throw new ClientError('_id is mandatory for a network.', EErrorsCodes.UNKNOWN_ERROR);
        }

        this._id = props._id;

        this.debug = debug.extend(this._id);

        this.import(props);
    }
```

Reply:
I'll update the debug assignment.  I'll also add some guards on the update() and save() to ensure the _id exists:
```
if (!this._id) {
    throw new Error...
```

If you're comfortable with the proposed changes, I'll look into the getter/setter and private class fields another time 😄 .  Looks like there's a bit to learn there.

Kyle

Reply:
@KyleBrandon no problem ( I know this is too much for the moment ;) ) . 

I wait your commit :)